### ISSUE TITLE
Freezing OP-TEE fails during suspend

### ISSUE THREAD

**Issue by cgellner** [OP-TEE/optee_os#4581]
When trying to suspend to RAM the OP-TEE TA user-space tasks using OP-TEE might fail to freeze

```
PM: suspend entry (s2idle)
Filesystems sync: 0.000 seconds
Freezing user space processes ... 
Freezing of tasks failed after 20.005 seconds (1 tasks refusing to freeze, wq_busy=0):
task:optee_example_s state:R  running task     stack:    0 pid:  128 ppid:     1 flags:0x00000001
[<807d3e24>] (__schedule) from [<c01bae20>] (0xc01bae20)
OOM killer enabled.
Restarting tasks ... done.
PM: suspend exit
```

According to my understanding this happens if a TA, triggered by the task to freeze, is trying to communicate to *tee-supplicant*, but *tee-supplicant* has been frozen first.

In order to reproduce OP-TEE QEMU environment can be used.
In order to increase the probability of the TA trying to communicate to *tee-supplicant* during suspend, the below patch should be applied to *secure_storage_ta.c* in *optee_examples*.

```diff --git a/secure_storage/ta/secure_storage_ta.c b/secure_storage/ta/secure_storage_ta.c
index edaf6a3..39f5ed2 100644
--- a/secure_storage/ta/secure_storage_ta.c
+++ b/secure_storage/ta/secure_storage_ta.c
@@ -256,11 +256,19 @@ TEE_Result TA_InvokeCommandEntryPoint(void __unused *session,
 				      uint32_t param_types,
 				      TEE_Param params[4])
 {
+	TEE_Result res;
+	size_t i;
+
 	switch (command) {
 	case TA_SECURE_STORAGE_CMD_WRITE_RAW:
 		return create_raw_object(param_types, params);
 	case TA_SECURE_STORAGE_CMD_READ_RAW:
-		return read_raw_object(param_types, params);
+		for (i=0; i<400; i++) {
+			res = read_raw_object(param_types, params);
+			if (res != TEE_SUCCESS)
+				break;
+		}
+		return res;
 	case TA_SECURE_STORAGE_CMD_DELETE:
 		return delete_object(param_types, params);
 	default:
--
```

Afterwards you can reproduce by executing
```(optee_example_secure_storage &); sleep 1; echo -n mem > /sys/power/state```
from non-secure world.

During debugging I could see, that while kernel trying to freeze the task is looping in
*drivers/tee/supp.c* - *optee_supp_thrd_req*
```while (wait_for_completion_interruptible(&req->c))```
Note: In my understanding other cases might be possible as well e.g.
- optee_cq_wait_for_completion waiting for available OP-TEE Thread
- one TA waiting in OPTEE_MSG_RPC_WAIT_QUEUE_SLEEP (*wq_sleep*) for other TA calling OPTEE_MSG_RPC_WAIT_QUEUE_WAKEUP (*wq_wakeup*). But second TA is blocked due to one of the other cases.

When I modified *kernel/power/process.c* - *try_to_freeze_tasks* - not to freeze *tee-supplicant* when freezing user tasks, the issue is not reproducible and suspend will happen after the TA did complete its work.
I wonder if there is a generic kernel mechanism to avoid freezing of user-space helper tasks like *tee-supplicant*.
How is this solved for fuse-FS? Isn't it using some user-space tasks for processing as well?

When I modified *drivers/tee/supp.c* - *optee_supp_thrd_req* - to call *try_to_freeze* in the *wait_for_completion_interruptible* loop mentioned above, the issue is not reproducible and suspend will happen although TA did not yet complete its work. TA will continue after the resume.
I'm not clear, if it is legal from OP-TEE perspective to suspend while TA is still executing.

**Comment by etienne-lms** [comment]
Thanks a lot for the description. Pretty clear and detailed.
This all makes sense. I think the last proposal is the right way.

**Comment by cgellner** [comment]
I proposed a set of patches via mailing lists 
```[RFC PATCH 0/3] tee: optee: Allow to freeze when tee-supplicant is frozen```

Currently waiting for some more feedback

**Comment by cgellner** [comment]
Beside the response of @jenswi-linaro I didn't see any feedback on RFC tagged patches

Therefore I send patches again - this time without RFC tag
`[PATCH 0/3] tee: optee: Allow to freeze when tee-supplicant is frozen`

**Comment by etienne-lms** [comment]
Can this thread be re-open and tagged **enhancement**?

**Comment by cgellner** [comment]
Hi,
I have resend the request to op-tee@lists.trustedfirmware.org; linux-kernel@vger.kernel.org last August.
- https://lkml.iu.edu/hypermail/linux/kernel/2108.1/00496.html
- https://lkml.iu.edu/hypermail/linux/kernel/2108.1/00498.html
- https://lkml.iu.edu/hypermail/linux/kernel/2108.1/00499.html
I didn't see any reaction at all.
Not 100% sure if I missed anything or if I did anything incorrectly.
Hints how I can improve are welcome.

Do you know if the mentioned "suspend" problem has been solved differently ?
I didn't test with most recent OP-TEE or kernel.

I have seen that latest kernel has some small changes (787c80cc7b22804aa370f04a19f9fe0fa98b1e49)
So I believe patch 3/3 can't be applied, but I assume the issue is still there in the new code (still using wait_for_completion)

In case there is some interest in this patch I could 
- reproduce issue again in latest version
- rework the patch set so it can be applied on latest kernel version
- resend the patch request

But before starting I would like to understand if there is some general interest in this patch at all or if failing to freeze during suspend is no problem from your perspective.

**Comment by jforissier** [comment]
Hi @cgellner,

Technically speaking I can't say if your patches are correct because I don't know the Linux suspend mechanisms well enough. However it seems you are addressing a real issue with these patches, so I think they deserve more attention.

> In case there is some interest in this patch I could
> 
>     * reproduce issue again in latest version
> 
>     * rework the patch set so it can be applied on latest kernel version
> 
>     * resend the patch request

@jenswi-linaro I think it is the way to go, isn't it? Maybe CC some experts of the "freeze" interface, too?

**Comment by jenswi-linaro** [comment]
I agree and the same problem here that I don't know that area well enough.

**Comment by victorchenpan** [comment]
I have meet the problem too. When tee-supplicant has been frozen, and then other task request tee-supplicant, it can't enter freezing flow.  So kernel can't finish freeze flow, and will not enter suspend flow.

Your patch can solve the problem.

**Comment by jforissier** [comment]
The last trace of @cgellner 's patchset I could see on the Linux Kernel Mailing list is https://lore.kernel.org/all/20210809071452.28508-1-cgellner@de.adit-jv.com/ (August 2021) and it has never gathered any Acked-by, Reviewed-by or Tested-by apparently (at least there is no such tags in the patches themselves). @victorchenpan are you subscribed to the LKML? Would you be able to provide feedback if the series was to be re-posted? Or perhaps even reply to the already sent emails by following the "Reply instructions" on the ML page?

**Comment by baiqiang728** [comment]
> Hi, I have resend the request to [op-tee@lists.trustedfirmware.org](mailto:op-tee@lists.trustedfirmware.org); [linux-kernel@vger.kernel.org](mailto:linux-kernel@vger.kernel.org) last August.
> 
> * https://lkml.iu.edu/hypermail/linux/kernel/2108.1/00496.html
> * https://lkml.iu.edu/hypermail/linux/kernel/2108.1/00498.html
> * https://lkml.iu.edu/hypermail/linux/kernel/2108.1/00499.html
>   I didn't see any reaction at all.
>   Not 100% sure if I missed anything or if I did anything incorrectly.
>   Hints how I can improve are welcome.
> 
> Do you know if the mentioned "suspend" problem has been solved differently ? I didn't test with most recent OP-TEE or kernel.
> 
> I have seen that latest kernel has some small changes (787c80cc7b22804aa370f04a19f9fe0fa98b1e49) So I believe patch 3/3 can't be applied, but I assume the issue is still there in the new code (still using wait_for_completion)
> 
> In case there is some interest in this patch I could
> 
> * reproduce issue again in latest version
> * rework the patch set so it can be applied on latest kernel version
> * resend the patch request
> 
> But before starting I would like to understand if there is some general interest in this patch at all or if failing to freeze during suspend is no problem from your perspective.

Thanks for your patch @cgellner ， I met the same issue.  After I update your patch, it seems that it works. However, if I do more test , such as stress test for STR when xtest is running. It may cause CPU rcu stall issue.  Do you have similar phenomena ？

Thanks

**Comment by LeiSen62** [comment]
- Kernel tee patches:
  - `tee: optee: prevent use-after-free when the client exits before the supplicant`
  - `tee: optee: Fix supplicant wait loop`
  The community is using these two patches to resolve the issue. However, since the task wait relies on wait_for_completion_killable, this function will return a non-zero value when a kill signal is received, causing optee-os to interpret the task as failed. Is this the expected behavior? Are there any other discussions regarding these two patches?
